File: programmers/etc/level2/다리를_지나는_트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridgeWeight = 0
    waiting = deque(truck_weights)
    bridge = deque([0]*bridge_length)
    
    while waiting or bridgeWeight > 0:
        bridgeWeight -= bridge.popleft()
        
        if waiting and bridgeWeight+waiting[0] <= weight:
            bridge.append(waiting[0])
            bridgeWeight += waiting.popleft()
        else:
            bridge.append(0)
        
        answer += 1
        
    return answer


bridge_length, weight = 2, 10
truck_weights = [7,4,5,6]
print(solution(bridge_length, weight, truck_weights)) # 8

bridge_length, weight = 100, 100
truck_weights = [10]
print(solution(bridge_length, weight, truck_weights)) # 101

bridge_length, weight = 100, 100
truck_weights = [10,10,10,10,10,10,10,10,10,10]
print(solution(bridge_length, weight, truck_weights)) # 110

File: Java/programmers/etc/level2/다리를지나는트럭.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 다리를지나는트럭 {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int bridgeWeight = 0;
        
        Queue<Integer> waiting = new LinkedList<>(){};
        for (int truck : truck_weights) {
            waiting.add(truck);
        }

        Queue<Integer> bridge = new LinkedList<>();
        for (int i=0; i<bridge_length; i++) {
            bridge.add(0);
        }

        while (!waiting.isEmpty() || bridgeWeight > 0) {
            bridgeWeight -= bridge.poll();

            if (!waiting.isEmpty() && bridgeWeight+waiting.peek() <= weight) {
                bridge.add(waiting.peek());
                bridgeWeight += waiting.poll();
            } else {
                bridge.add(0);
            }

            answer++;
        }

        return answer;
    }
}
